tokio/src/task/scope/scope.rs

```diff
@@ -0,0 +1,354 @@
+    }
+}
+
+pin_project! {
+    /// Allows to wait for a child task to join
+    pub struct ScopedJoinHandle<'scope, T> {
+        #[pin]
+        handle: JoinHandle<Result<T, CancellableFutureError>>,
+        phantom: core::marker::PhantomData<&'scope ()>,
+    }
+}
+
+impl<'scope, T> Future for ScopedJoinHandle<'scope, T> {
+    // The actual type is Result<Result<T, CancellableFutureError>, JoinError>
+    // However the cancellation will only happen at the exit of the scope. This
+    // means in all cases the user still has a handle to the task, the task can
+    // not be cancelled yet.
+    type Output = Result<T, JoinError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.project()
+            .handle
+            .poll(cx)
+            .map(|poll_res| poll_res.map(|poll_ok| poll_ok.expect("Task can not be cancelled")))
+    }
+}
+
+/// A handle to the scope, which allows to spawn child tasks
+#[derive(Clone)]
+pub struct ScopeHandle {
+    scope: ScopeState,
+}
+
+impl core::fmt::Debug for ScopeHandle {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("ScopeHandle").finish()
+    }
+}
+
+struct CancelTasksGuard<'a> {
+    scope: &'a CancellationToken,
+}
+
+impl<'a> Drop for CancelTasksGuard<'a> {
+    fn drop(&mut self) {
+        self.scope.cancel();
+    }
+}
+
+struct WaitForTasksToJoinGuard<'a> {
+    wait_group: &'a SharedWaitGroup,
+    drop_behavior: ScopeDropBehavior,
+    enabled: bool,
+}
+
+impl<'a> WaitForTasksToJoinGuard<'a> {
+    fn disarm(&mut self) {
+        self.enabled = false;
+    }
+}
+
+impl<'a> Drop for WaitForTasksToJoinGuard<'a> {
+    fn drop(&mut self) {
+        if !self.enabled {
+            return;
+        }
+
+        match self.drop_behavior {
+            ScopeDropBehavior::BlockToCompletion => {
+                let wait_fut = self.wait_group.wait_future();
+
+                // TODOs:
+                // - This should not have a futures dependency
+                // - This might block multithreaded runtimes, since the tasks might need
+                //   the current executor thread to make progress, due to dependening on
+                //   its IO handles. We need to do something along task::block_in_place
+                //   to solve this.
+                futures::executor::block_on(wait_fut);
+            }
+            ScopeDropBehavior::Panic => {
+                panic!("Scope was dropped before child tasks run to completion");
+            }
+            ScopeDropBehavior::Abort => {
+                eprintln!("[ERROR] A scope was dropped without being awaited");
+                std::process::abort();
+            }
+            ScopeDropBehavior::ContinueTasks => {
+                // Do nothing
+            }
+        }
+    }
+}
+
+impl ScopeHandle {
+    /// spawns a task on the scope
+    pub fn spawn<'inner, T, R>(&'inner self, task: T) -> ScopedJoinHandle<'inner, R>
```

Comment:
For ergonomics reasons, I think we probably ought to have a free-fn `spawn` (or `spawn_scoped` if it's exported at the top level?) that's available inside the scope, similar to `spawn_local`...

Reply:
The non-free function has an advantage: It also constrains the lifetime of the join handles to the ones of the `ScopeHandle`. Since the `ScopeHandle` doesn't allow to spawn outside of the scope, the `ScopedJoinHandle` can thereby also never resolve to a cancelled variant. If we would make this a free function there could either be no current scope available, or the scope might already have been cancelled.

Another thing is that `ScopeHandle`s can currently be cloned and stored in structs for spawning later - at least as long as that struct lives within the lifetime of a scope.

Reply:
> If we would make this a free function there could either be no current scope available, or the scope might already have been cancelled.

I think it would be fine for a free function to panic in these cases; that's what the rest of `tokio` does...

> It also constrains the lifetime of the join handles to the ones of the `ScopeHandle`.

_This_, on the other hand, is a compelling reason to use the handle only. AFAICT, there isn't really any other way to constrain the scoped join handle lifetime. I think the ergonomics of a free function would be better, but I think the ability to return a `JoinHandle` that only lives as long as the scope is super valuable, so this may be the best approach!